### PR TITLE
Code changes for running SGX inside Docker

### PR DESCRIPTION
**Pull request by manju956**
A new docker-compose file is added for running SGX in HW mode. 
By default, PoET runs with SGX_SIM mode. 

Signed-off-by: manju.ac <manjunath.a.c@intel.com>

### CODE REVIEW

File: build_sgx_binary.sh [hunk 1]
```diff
@@ -1,15 +1,23 @@
 #!/bin/bash
-WORK_DIR=/project/sawtooth-poet/src
+WORK_DIR=/project/sawtooth-poet2/src
```
Comment (manojgop) [inline build_sgx_binary.sh]: can we use sawtooth-poet

File: build_sgx_binary.sh [hunk 2]
```diff
@@ -1,15 +1,23 @@
...
 SGX_DIR=/tmp/sgxsdk
 
 #setting SGX environment
 source $SGX_DIR/environment
 
+#Add proxy settings
```
Comment (manojgop) [inline build_sgx_binary.sh]: Proxy and aesm can be in an if block if SGX is hw mode

File: docker-compose-hw.yaml [hunk 3]
```diff
@@ -0,0 +1,118 @@
+# Copyright 2018 Intel Corporation
```
Comment (manojgop) [inline docker-compose-hw.yaml]: Rename the fine as docker-compose-sgx-hw.yaml

File: docker-compose-hw.yaml [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+  poet-2-engine-1:
+    build:
+      context: .
+      dockerfile: poet2-sgx-engine.dockerfile
+      args:
+        - http_proxy
+        - https_proxy
+        - no_proxy
+    volumes:
+      - ./:/project/sawtooth-poet2/
```
Comment (manojgop) [inline docker-compose-hw.yaml]: use /project/sawtooth-poet

File: build_sgx_binary.sh (hunk 1 above)
Comment (manju956) [reply build_sgx_binary.sh]: Ok, will modify to sawtooth-poet

File: build_sgx_binary.sh (hunk 2 above)
Comment (manju956) [reply build_sgx_binary.sh]: Ok

File: docker-compose-hw.yaml (hunk 3 above)
Comment (manju956) [reply docker-compose-hw.yaml]: Ok

File: docker-compose-hw.yaml (hunk 4 above)
Comment (manju956) [reply docker-compose-hw.yaml]: Ok

File: build_sgx_binary.sh [hunk 5]
```diff
@@ -5,11 +5,26 @@ SGX_DIR=/tmp/sgxsdk
...
+
 #building SGX bridge and Enclave
 cd $WORK_DIR
 mkdir build
 cd build
-cmake $WORK_DIR/sgx
+if [ $SGX_HW_MODE ]; then
+    SGX_HW_MODE=$SGX_HW_MODE cmake $WORK_DIR/sgx
+else
+    cmake $WORK_DIR/sgx
```
Comment (manojgop) [inline build_sgx_binary.sh]: Add comment it is simulator by default

File: src/sgx/CMakeLists.txt [hunk 6]
```diff
@@ -35,9 +35,13 @@ SET(TEST_POET_ENCLAVE testEnclave)
 # ADDITIONAL LIBRARIES
 # XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
 
-# SGX -- should move to findsgx 
-OPTION(SGX_USE_SIMULATOR "Use the sgx simulator" TRUE)
-
+# SGX -- should move to findsgx
```
Comment (manojgop) [inline src/sgx/CMakeLists.txt]: can we remove this comment

Comment (manju956) [reply src/sgx/CMakeLists.txt]: Ok

File: build_sgx_binary.sh (hunk 5 above)
Comment (manju956) [reply build_sgx_binary.sh]: Ok

File: src/validator-registry-tp/src/validator_registry_tp.rs [hunk 7]
```diff
@@ -156,8 +155,8 @@ impl ValidatorRegistryTransactionHandler {
         // registry
 
         let mut validator_info_address: String;
-
-        for entry_str in validator_map.entries.iter_mut() {
+        for idx in 0..validator_map.entries.len() {
+            let mut entry_str = validator_map.entries.get_mut(idx).unwrap();
```
[the remark this answers is not in the document]
Comment (manju956) [reply src/validator-registry-tp/src/validator_registry_tp.rs]: Expect would also crash, I will add a message saying it's unexpected to end up like this.

File: src/validator-registry-tp/src/validator_registry_tp.rs [hunk 8]
```diff
@@ -322,28 +325,44 @@ impl ValidatorRegistryTransactionHandler {
...
     fn _get_state(&self, context: &mut TransactionContext,
-                  address: &String, ) -> String {
-        let entries_ = context.get_state(vec![address.to_string()]); // this return Vec<u8>
+                  address: &String, ) -> Result<String, String> {
+        let entries_ = context.get_state(vec![address.to_string()]); // this return Option<Vec<u8>>
         let entries = if entries_.is_ok() {
             entries_.expect("Error reading entries")
         } else {
             warn!("Could not get context for address : {}", address);
             panic!("Error getting context.");
```
[the remark this answers is not in the document]
Comment (manju956) [reply src/validator-registry-tp/src/validator_registry_tp.rs]: Ok

**Pull request merged**